### PR TITLE
add confirm dialog for admin config update action

### PR DESCRIPTION
**Pull request by kiranparajuli589**
#### Description
- implement confirm dialog to update admin credentials
- delete user prefs. if updated admin config is not ok
- added button to submit the form

#### Related
- [OP#42273] part of https://community.openproject.org/projects/nextcloud-integration/work_packages/42273

### CODE REVIEW

File: lib/Controller/ConfigController.php [hunk 1]
```diff
@@ -115,7 +134,13 @@ public function setAdminConfig(array $values): DataResponse {
 		foreach ($values as $key => $value) {
 			$this->config->setAppValue(Application::APP_ID, $key, trim($value));
 		}
-		return new DataResponse(1);
+		// if the admin config is not ok, we clear every existing user information
```
Comment (individual-it) [inline lib/Controller/ConfigController.php:137]: we clear it in any case

Comment (individual-it) [inline lib/Controller/ConfigController.php:137]: we clear it in any case

File: src/components/AdminSettings.vue [hunk 2]
```diff
@@ -65,17 +68,35 @@ export default {
...
+		updateForm() {
 			const that = this
-			delay(() => {
-				that.saveOptions()
-			}, 2000)()
+			OC.dialogs.confirmDestructive(
+				t(
+					'integration_openproject',
+					'Are you sure you want to update the admin settings?'
+					+ ' After saving, every connected users must need to re-connect to the Openproject instance.'
```
Comment (individual-it) [inline src/components/AdminSettings.vue:82]:
```suggestion
					+ ' After saving, every connected user will have to re-connect to the OpenProject instance.'
```

File: src/components/AdminSettings.vue [hunk 3]
```diff
@@ -65,17 +68,35 @@ export default {
...
-		onInput() {
+		updateForm() {
 			const that = this
-			delay(() => {
-				that.saveOptions()
-			}, 2000)()
+			OC.dialogs.confirmDestructive(
+				t(
+					'integration_openproject',
+					'Are you sure you want to update the admin settings?'
```
Comment (individual-it) [inline src/components/AdminSettings.vue:81]:
```suggestion
					'Are you sure you want to update the Oauth settings for OpenProject?'
```

File: tests/jest/components/AdminSettings.spec.js [hunk 4]
```diff
@@ -40,29 +61,99 @@ describe('AdminSettings', () => {
...
+				confirmSpy = jest.spyOn(global.OC.dialogs, 'confirmDestructive')
+				wrapper = getWrapper({
+					isAdminConfigOk: false,
 				})
-			const wrapper = getWrapper()
-			const inputField = wrapper.find(inputSelector)
-
-			await inputField.trigger('input')
+			})
+			it('should show the save button', async () => {
```
Comment (individual-it) [inline tests/jest/components/AdminSettings.spec.js:76]:
```suggestion
			it('should show the update button', async () => {
```

File: tests/jest/components/AdminSettings.spec.js [hunk 5]
```diff
@@ -40,29 +61,99 @@ describe('AdminSettings', () => {
...
+				expect(wrapper.find(selectors.saveConfigButton)).toMatchSnapshot()
+				expect(wrapper.find(selectors.updateConfigButton).exists()).toBeFalsy()
```
Comment (individual-it) [inline tests/jest/components/AdminSettings.spec.js:78]: I think this should be true

File: tests/jest/components/AdminSettings.spec.js [hunk 6]
```diff
@@ -40,29 +61,99 @@ describe('AdminSettings', () => {
...
 			jest.clearAllMocks()
 		})
-		it.each([
-			selectors.oauthInstance,
-			selectors.oauthClientId,
-			selectors.oauthClientId,
-		])('should call "onInput" method', async (inputSelector) => {
-			const onInputSpy = jest.spyOn(AdminSettings.methods, 'onInput')
-				.mockImplementationOnce(() => {
+		describe('when the admin config status is not ok beforehand', () => {
```
Comment (individual-it) [inline tests/jest/components/AdminSettings.spec.js:68]:
```suggestion
		describe('admin config is not complete', () => {
```

File: tests/jest/components/AdminSettings.spec.js [hunk 7]
```diff
@@ -40,29 +61,99 @@ describe('AdminSettings', () => {
...
+			})
+			it('should not trigger confirm dialog on save', async () => {
+				const saveConfigButton = wrapper.find(selectors.saveConfigButton)
+				const inputField = wrapper.find(selectors.oauthClientId)
+				await inputField.setValue('test')
+				await saveConfigButton.trigger('click')
+				expect(confirmSpy).toBeCalledTimes(0)
+			})
+		})
+		describe('when the admin config status is ok beforehand', () => {
```
Comment (individual-it) [inline tests/jest/components/AdminSettings.spec.js:88]:
```suggestion
		describe('admin config is complete', () => {
```

File: tests/jest/components/AdminSettings.spec.js [hunk 8]
```diff
@@ -40,29 +61,99 @@ describe('AdminSettings', () => {
...
+			let wrapper, confirmSpy
+			beforeEach(() => {
+				confirmSpy = jest.spyOn(global.OC.dialogs, 'confirmDestructive')
+				wrapper = getWrapper({
+					isAdminConfigOk: true,
+				})
+			})
+			it('should show the save button', async () => {
+				expect(wrapper.find(selectors.updateConfigButton)).toMatchSnapshot()
+				expect(wrapper.find(selectors.saveConfigButton).exists()).toBeFalsy()
```
Comment (individual-it) [inline tests/jest/components/AdminSettings.spec.js:98]: shouldn't that be true?

File: tests/lib/Controller/ConfigControllerTest.php [hunk 9]
```diff
@@ -286,4 +298,107 @@ public function testOauthNoAccessTokenInResponse($oauthResponse, $expectedRedire
...
+	}
+
+	/**
+	 * @param array<string> $credsToUpdate
+	 * @param bool $adminConfigStatus
+	 * @dataProvider setAdminConfigStatusDataProvider
+	 * @return void
+	 */
+	public function testSetAdminConfigForDifferentAdminConfigStatus($credsToUpdate, $adminConfigStatus) {
+		$manager = \OC::$server->getUserManager();
```
Comment (individual-it) [inline tests/lib/Controller/ConfigControllerTest.php:332]:
```suggestion
		$userManager = \OC::$server->getUserManager();
```

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/lib/Controller/ConfigControllerTest.php [hunk 10]
```diff
@@ -286,4 +298,107 @@ public function testOauthNoAccessTokenInResponse($oauthResponse, $expectedRedire
...
+		$count = 0;
+		$function = function (IUser $user) use (&$count) {
+			$count++;
+		};
+		$manager->callForAllUsers($function, '', true);
+		$this->assertSame(1, $count, 'Expected to have only 1 user in the dB before this test');
```
Comment (individual-it) [inline tests/lib/Controller/ConfigControllerTest.php:338]: why should we fail tests if there are more then one user before?

Comment (kiranparajuli589) [reply tests/lib/Controller/ConfigControllerTest.php:339]: I've to mock the `deleteUserValue` fn for the number of fn calls and consecutive args to verify cleanup is done for every user. The consecutive args `withConsecutive` can't be fed calculated data, so I have to have only two to match with the CI. One will be an admin and another will be the test-created user.

File: tests/jest/components/AdminSettings.spec.js (hunk 4 above)
Comment (kiranparajuli589) [reply tests/jest/components/AdminSettings.spec.js:76]: when the admin config is incomplete, it shows the save btn without the confirm dialog. isn't that how it should behave?

File: tests/jest/components/AdminSettings.spec.js (hunk 5 above)
Comment (kiranparajuli589) [reply tests/jest/components/AdminSettings.spec.js:78]: if the admin config is incomplete, only the save btn is displayed

File: tests/jest/components/AdminSettings.spec.js (hunk 8 above)
Comment (kiranparajuli589) [reply tests/jest/components/AdminSettings.spec.js:98]: if the admin config is complete, then only update btn is displayed which then triggers the confirm dialog. isn't this how it should be?

File: tests/lib/Controller/ConfigControllerTest.php (hunk 10 above)
Comment (kiranparajuli589) [reply tests/lib/Controller/ConfigControllerTest.php:339]: if we can provide the args like this then the check wouldn't have been required. can we do it like so?
```
->withConsecutive(listOfConsecutiveArgs)
```

File: src/components/AdminSettings.vue [hunk 11]
```diff
@@ -65,17 +68,35 @@ export default {
...
+					'Are you sure you want to update the Oauth settings for OpenProject?'
+					+ ' After saving, all connected users will have to re-connect to the OpenProject instance.'
+				),
+				t('integration_openproject', 'Confirm Update'),
+				{
+					type: OC.dialogs.YES_NO_BUTTONS,
+					confirm: t('integration_openproject', 'Update'),
+					confirmClasses: 'error',
+					cancel: t('integration_openproject', 'Cancel'),
+				},
```
Comment (individual-it) [inline src/components/AdminSettings.vue:90]: replace with suggestion of @wielinde in https://community.openproject.org/projects/nextcloud-integration/work_packages/42273/activity#activity-4

**Review by individual-it — CHANGES_REQUESTED** [review]

**Review by individual-it — APPROVED** [review]

**Comment by individual-it** [comment]
this should also fix https://community.openproject.org/projects/nextcloud-integration/work_packages/42272 or?

**Comment by kiranparajuli589** [comment]
> this should also fix https://community.openproject.org/projects/nextcloud-integration/work_packages/42272 or?

yes it should

**Pull request merged**